### PR TITLE
Updated math.py & nn.py with docstrings.

### PR DESCRIPTION
**Pull request by Sheiphan**
Updated `keras_core/ops/math.py` and `keras_core/ops/nn.py` with docstring.

Please do let me know if any changes are needed.

### CODE REVIEW

File: keras_core/ops/math.py [hunk 1]
```diff
@@ -63,6 +84,30 @@ def call(self, x):
...
+        x: Input tensor.
+        k: An integer representing the number of top elements to retrieve.
+        sorted: A boolean indicating whether to sort the output in descending order.
+            Default is True.
+
+    Returns:
+        A tuple containing two tensors. The first tensor contains the top-k values, and
+        the second tensor contains the indices of the top-k values in the input tensor.
+
+    Raises:
```
Comment (fchollet) [inline keras_core/ops/math.py:100]: All raises sections should be removed.

File: keras_core/ops/math.py [hunk 2]
```diff
@@ -35,6 +35,27 @@ def call(self, data, segment_ids):
...
+        num_segments: An integer representing the total number of segments. If not specified,
+            it is inferred from the maximum value in `segment_ids`.
+        sorted: A boolean indicating whether `segment_ids` is sorted. Default is False.
+
+    Returns:
+        A tensor containing the sum of segments, where each element represents the sum of
+        the corresponding segment in `data`.
+
+    Example:
+        >>> data = tf.constant([1, 2, 3, 4, 5, 6])
```
Comment (fchollet) [inline keras_core/ops/math.py:53]: All code examples should use Keras ops, never TF tensors.

File: keras_core/ops/math.py [hunk 3]
```diff
@@ -35,6 +35,27 @@ def call(self, data, segment_ids):
...
+        segment_ids: A 1-D tensor containing segment indices for each element in `data`.
+        num_segments: An integer representing the total number of segments. If not specified,
+            it is inferred from the maximum value in `segment_ids`.
+        sorted: A boolean indicating whether `segment_ids` is sorted. Default is False.
+
+    Returns:
+        A tensor containing the sum of segments, where each element represents the sum of
+        the corresponding segment in `data`.
+
+    Example:
```
Comment (fchollet) [inline keras_core/ops/math.py:52]: Examples should be formatted as code blocks.

File: keras_core/ops/math.py [hunk 4]
```diff
@@ -35,6 +35,27 @@ def call(self, data, segment_ids):
...
 def segment_sum(data, segment_ids, num_segments=None, sorted=False):
+    """
+    Computes the sum of segments in a tensor.
+
+    Args:
+        data: Input tensor.
+        segment_ids: A 1-D tensor containing segment indices for each element in `data`.
+        num_segments: An integer representing the total number of segments. If not specified,
+            it is inferred from the maximum value in `segment_ids`.
+        sorted: A boolean indicating whether `segment_ids` is sorted. Default is False.
```
Comment (fchollet) [inline keras_core/ops/math.py:46]: All keywords such as `False` should have backticks.

File: keras_core/ops/math.py [hunk 5]
```diff
@@ -35,6 +35,27 @@ def call(self, data, segment_ids):
 
 @keras_core_export("keras_core.ops.segment_sum")
 def segment_sum(data, segment_ids, num_segments=None, sorted=False):
+    """
+    Computes the sum of segments in a tensor.
```
Comment (fchollet) [inline keras_core/ops/math.py:39]: The one-liner description should always be on the first line (above)

File: keras_core/ops/math.py [hunk 6]
```diff
@@ -103,6 +172,25 @@ def call(self, x):
 
 @keras_core_export("keras_core.ops.logsumexp")
 def logsumexp(x, axis=None, keepdims=False):
+    """
+    Computes the logarithm of the sum of exponentials of elements in a tensor.
+
+    Args:
+        x: Input tensor.
+        axis: An integer or a tuple of integers specifying the axis/axes along which to compute
```
Comment (fchollet) [inline keras_core/ops/math.py:180]: Make sure that all lines stay under 80 char.

File: keras_core/ops/nn.py [hunk 7]
```diff
@@ -959,6 +959,39 @@ def compute_output_spec(self, x):
...
+    """
+    Converts integer tensor `x` into a one-hot tensor.
+
+    The one-hot encoding is a representation where each integer value is
+    converted into a binary vector with a length equal to `num_classes`,
+    and the index corresponding to the integer value is marked as 1, while
+    all other indices are marked as 0.
+
+    Args:
+        x (KerasTensor): Integer tensor to be encoded. The shape can be
```
Comment (fchollet) [inline keras_core/ops/nn.py:971]: Rather than specifying the type before `:`, specify it after, like `x: Integer.`

Note that this input is NOT a KerasTensor. "Integer tensor" is the right description.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by Sheiphan** [comment]
I feel like I have made all the necessary changes. 
Do let me know if anything needs to be changed.

File: keras_core/ops/math.py [hunk 8]
```diff
@@ -35,6 +35,31 @@ def call(self, data, segment_ids):
...
+        A tensor containing the sum of segments, where each element 
+        represents the sum of the corresponding segment in `data`.
+
+    Example:
+    ```
+    >>> data = keras_core.ops.convert_to_tensor([1, 2, 3, 4, 5, 6])
+    >>> segment_ids = keras_core.ops.convert_to_tensor([0, 1, 0, 1, 0, 1])
+    >>> result = segment_sum(data, segment_ids)
+    >>> print(result)
+    tf.Tensor([9 12], shape=(2,), dtype=int32)
```
Comment (fchollet) [inline keras_core/ops/math.py:60]: Just print the values, no references to TF. E.g. you can just print `array([9 12], shape=(2,), dtype=int32)`

File: keras_core/ops/math.py [hunk 9]
```diff
@@ -63,6 +88,29 @@ def call(self, x):
 
 @keras_core_export("keras_core.ops.top_k")
 def top_k(x, k, sorted=True):
+    """Finds the top-k values and their indices in a tensor.
+
+    Args:
+        x: Input tensor.
+        k: An integer representing the number of top elements to retrieve.
+        sorted: A boolean indicating whether to sort the output in 
+            descending order.Default is `True`.
```
Comment (fchollet) [inline keras_core/ops/math.py:97]: Add space after `.`

File: keras_core/ops/math.py [hunk 10]
```diff
@@ -35,6 +35,31 @@ def call(self, data, segment_ids):
...
+    Returns:
+        A tensor containing the sum of segments, where each element 
+        represents the sum of the corresponding segment in `data`.
+
+    Example:
+    ```
+    >>> data = keras_core.ops.convert_to_tensor([1, 2, 3, 4, 5, 6])
+    >>> segment_ids = keras_core.ops.convert_to_tensor([0, 1, 0, 1, 0, 1])
+    >>> result = segment_sum(data, segment_ids)
+    >>> print(result)
```
Comment (fchollet) [inline keras_core/ops/math.py:59]: You don't need the lines `print(result)`. You can just do e.g.

```
>>> segment_sum(data, segment_ids)
array([9 12], shape=(2,), dtype=int32)
```

File: keras_core/ops/math.py [hunk 11]
```diff
@@ -103,6 +173,27 @@ def call(self, x):
 
 @keras_core_export("keras_core.ops.logsumexp")
 def logsumexp(x, axis=None, keepdims=False):
+    """Computes the logarithm of the sum of exponentials of elements in a tensor.
+
+    Args:
+        x: Input tensor.
+        axis: An integer or a tuple of integers specifying the axis/axes along 
+            which to compute the sum. If None, the sum is computed over all elements.
```
Comment (fchollet) [inline keras_core/ops/math.py:181]: None is a code keyword and should have backticks

File: keras_core/ops/nn.py [hunk 12]
```diff
@@ -993,6 +1028,39 @@ def compute_output_spec(self, target, output):
...
 def binary_crossentropy(target, output, from_logits=False):
+    """Computes the binary cross-entropy loss between target tensor and output tensor.
+
+    The binary cross-entropy loss is commonly used in binary classification tasks
+    where each input sample belongs to one of the two classes. It measures the
+    dissimilarity between the target and output probabilities or logits.
+
+    Args:
+        target: The target tensor representing the true binary labels.
+                Its shape should match the shape of the `output` tensor.
```
Comment (fchollet) [inline keras_core/ops/nn.py:1039]: Indent with 4 spaces, no more

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

File: keras_core/ops/math.py [hunk 13]
```diff
@@ -35,6 +35,30 @@ def call(self, data, segment_ids):
 
 @keras_core_export("keras_core.ops.segment_sum")
 def segment_sum(data, segment_ids, num_segments=None, sorted=False):
+    """Computes the sum of segments in a tensor.
+
+    Args:
+        data: Input tensor.
+        segment_ids: A 1-D tensor containing segment indices for each 
+                element in `data`.
```
Comment (fchollet) [inline keras_core/ops/math.py:43]: Please use 4-space indent.

File: keras_core/ops/math.py [hunk 14]
```diff
@@ -35,6 +35,30 @@ def call(self, data, segment_ids):
...
+        num_segments: An integer representing the total number of 
+            segments. If not specified, it is inferred from the maximum
+            value in `segment_ids`.
+        sorted: A boolean indicating whether `segment_ids` is sorted. 
+                Default is `False`.
```
Comment (fchollet) [inline keras_core/ops/math.py:48]: Here too

File: keras_core/ops/math.py [hunk 15]
```diff
@@ -152,6 +240,30 @@ def call(self, x):
...
+            - 'reduced': Returns the reduced QR decomposition. (default)
+            - 'complete': Returns the complete QR decomposition.
+
+    Returns:
+        A tuple containing two tensors. The first tensor represents the 
+        orthogonal matrix Q, and the second tensor represents the upper 
+        triangular matrix R.
+
+    Example:
+    ```
```
Comment (fchollet) [inline keras_core/ops/math.py:257]: Add `python` at the start of the code blocks

File: keras_core/ops/math.py [hunk 16]
```diff
@@ -152,6 +240,30 @@ def call(self, x):
...
+    >>> x = keras_core.ops.convert_to_tensor([[1., 2.], [3., 4.], [5., 6.]])
+    >>> q, r = qr(x)
+    >>> print(q)
+    ([[-0.16903079  0.897085  ]
```
Comment (fchollet) [inline keras_core/ops/math.py:261]: Prefix with `array`, e.g. 

```
array([[-0.16903079  0.897085  ]
       [-0.5070925   0.2760267 ]
       [-0.8451542  -0.34503305]], shape=(3, 2), dtype=float32)
```

File: keras_core/ops/nn.py [hunk 17]
```diff
@@ -43,15 +43,15 @@
 
 class Relu(Operation):
     def call(self, x):
-        return backend.nn.relu(x)
+        return backend.nn.relu(x)
```
Comment (fchollet) [inline keras_core/ops/nn.py:46]: Make sure to run `format.sh`, it will remove extra whitespace and so on

File: keras_core/ops/nn.py [hunk 18]
```diff
@@ -993,6 +1026,38 @@ def compute_output_spec(self, target, output):
...
+            Default is `False`.
+
+    Returns:
+        Integer tensor: The computed binary cross-entropy loss between 
+        `target` and `output`.
+
+    Example:
+    ```
+    >>> target = keras_core.ops.convert_to_tensor([0, 1, 1, 0], 
+                                                  dtype=tf.float32)
```
Comment (fchollet) [inline keras_core/ops/nn.py:1054]: No reference to TF -- use `float32`

Note that you don't actually need to pass anything, the default is float32.

File: keras_core/ops/nn.py [hunk 19]
```diff
@@ -1082,6 +1186,45 @@ def compute_output_spec(self, target, output):
     ]
 )
 def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
+    """Computes the sparse categorical cross-entropy loss between target tensor and output tensor.
```
Comment (fchollet) [inline keras_core/ops/nn.py:1189]: Make sure to keep lines under 80 chars.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

**Comment by Sheiphan** [comment]
Thank you for your update.
```python
    >>> x = keras_core.ops.convert_to_tensor([1., 2., 3.])
    >>> logsumexp(x)
    (3.407606, shape=(), dtype=float32)
 ```
Should I add `array(3.407606, shape=(), dtype=float32)` as its just one number.

**Comment by Sheiphan** [comment]
Thanks you for your patience.
Check if everything is alright.

File: keras_core/ops/math.py [hunk 20]
```diff
@@ -63,6 +87,29 @@ def call(self, x):
 
 @keras_core_export("keras_core.ops.top_k")
 def top_k(x, k, sorted=True):
+    """Finds the top-k values and their indices in a tensor.
+
+    Args:
+        x: Input tensor.
+        k: An integer representing the number of top elements to retrieve.
+        sorted: A boolean indicating whether to sort the output in
+        descending order.Default is `True`.
```
Comment (fchollet) [inline keras_core/ops/math.py:96]: Add space after `.`

File: keras_core/ops/math.py [hunk 21]
```diff
@@ -82,6 +129,28 @@ def call(self, targets, predictions):
...
+        targets: A tensor of true labels.
+        predictions: A tensor of predicted labels.
+        k: An integer representing the number of predictions to consider.
+
+    Returns:
+        A boolean tensor of the same shape as `targets`, where each element
+        indicates whether the corresponding target is in the top-k predictions.
+        
+    Example:
+    ```python
```
Comment (fchollet) [inline keras_core/ops/math.py:144]: So, there are two ways to add code examples:

1. Using a python block, with 3 backticks, `python`, a line break, the code, a line break, then 3 backticks to close
2. Using `>>>` at the start of one or more lines

You should use one **or** the other -- here you can just remove the above line (same elsewhere)

File: keras_core/ops/math.py [hunk 22]
```diff
@@ -103,6 +172,27 @@ def call(self, x):
...
 def logsumexp(x, axis=None, keepdims=False):
+    """Computes the logarithm of sum of exponentials of elements in a tensor.
+
+    Args:
+        x: Input tensor.
+        axis: An integer or a tuple of integers specifying the axis/axes
+            along which to compute the sum. If `None`, the sum is computed
+            over all elements. Default is `None`.
+        keepdims: A boolean indicating whether to keep the dimensions of
+            the input tensor when computing the sum. Default is False.
```
Comment (fchollet) [inline keras_core/ops/math.py:183]: Backticks around False

File: keras_core/ops/nn.py [hunk 23]
```diff
@@ -1082,6 +1188,45 @@ def compute_output_spec(self, target, output):
...
+            Default is -1, which corresponds to the last dimension
+            of the tensors.
+
+    Returns:
+        Integer tensor: The computed sparse categorical cross-entropy
+        loss between `target` and `output`.
+        
+    Example:
+    ```python
+    >>> target = keras_core.ops.convert_to_tensor([0, 1, 2], dtype=tf.int32)
```
Comment (fchollet) [inline keras_core/ops/nn.py:1222]: Remove references to TF, just use strings like "int32"

File: keras_core/ops/math.py (hunk 21 above)
Comment (Sheiphan) [reply keras_core/ops/math.py:144]:
```python
    data = keras_core.ops.convert_to_tensor([1, 2, 3, 4, 5, 6])
    segment_ids = keras_core.ops.convert_to_tensor([0, 1, 0, 1, 0, 1])
    segment_sum(data, segment_ids)
    array([9 12], shape=(2,), dtype=int32)
```
Will this do?

**Review by fchollet — APPROVED** [review]
LGTM, thanks!
